tests/test_cmdline_options.py

```diff
@@ -0,0 +1,185 @@
+        min_size=0,
+        max_size=MAX_NUMER_OF_CODE_ARGS,
+    )
+    # there is at most one "-" entry in `code` (so only one stdin)
+    .filter(lambda x: len([c for c in x if c == CODE_STDIN_MARKER]) <= 1).map(
+        lambda xs: ["--code"] + xs if xs else []
+    )
+)
+
+deps_option_strategy = (
+    st.lists(
+        st.sampled_from(available_deps),
+        min_size=0,
+        max_size=MAX_NUMER_OF_DEPS_ARGS,
+        # TODO this is a temporary fix, to always result in return code = 0
+        # This test is supposed to give a "happy path"
+    )
+    .filter(lambda x: str(project_with_no_issues) in x or len(x) == 0)
+    .map(lambda xs: ["--deps"] + xs if xs else [])
+)
+
+ignored_strategy = st.lists(safe_string, min_size=0, max_size=MAX_IGNORE_ARGS)
+ignored_undeclared_strategy = ignored_strategy.map(
+    lambda xs: ["--ignore-undeclared"] + xs if xs else []
+)
+ignored_unused_strategy = ignored_strategy.map(
+    lambda xs: ["--ignore-unused"] + xs if xs else []
+)
+
+deps_parser_choice_strategy = st.one_of(
+    st.sampled_from(deps_parser_choice), st.none()
+).map(lambda x: ["--deps-parser-choice", x] if x is not None else [])
+
+verbosity_indicator = st.text(
+    alphabet=["q", "v"], min_size=1, max_size=MAX_VERBOSITY_ARGS
+)
+verbosity_strategy = st.lists(
+    verbosity_indicator, min_size=0, max_size=MAX_VERBOSITY_ARGS
+).map(lambda xs: [f"-{x}" for x in xs])
+
+venv_strategy = st.one_of(st.sampled_from(venvs), st.none()).map(
+    lambda x: ["--venv", x] if x is not None else []
+)
+
+
+# ---------------------------------------------- #
+# Compose available strategies
+# ---------------------------------------------- #
+@st.composite
+def cli_arguments_combinations(draw):
+    to_stdin = None
+    strategy = st.permutations(
+        [
+            draw(actions_strategy),
+            draw(output_formats_strategy),
+            draw(deps_option_strategy),
+            draw(ignored_undeclared_strategy),
+            draw(ignored_unused_strategy),
+            draw(deps_parser_choice_strategy),
+            draw(verbosity_strategy),
+            # draw(venv_strategy), commented out due to
+            #                      high execution time and planned changes of `--venv`
+        ]
+    )
+    args = reduce(lambda x, y: x + y, draw(strategy))
+
+    # only `code` option changes to_stdin
+    code = draw(code_option_strategy)
+    if code is not None:
+        if "-" in code:
+            to_stdin = example_python_stdin
+        args += code
+
+    return (args, to_stdin)
+
+
+@given(cli_arguments=cli_arguments_combinations())
+def test_options_interactions__correct_options__give_success_code(cli_arguments):
+    """Check if a combination of valid options
+
+    makes a valid run of fawltydeps CLI tool.
+    """
+    drawn_args, to_stdin = cli_arguments
+    args = [str(project_with_no_issues)] + drawn_args
+
+    _, _, returncode = run_fawltydeps(*args, to_stdin=to_stdin)
+
+    assert returncode == 0
```

Comment:
I have to admit a certain amount of 🤯 and my eyes glazed over somewhere around `cli_arguments_combinations()`, but there's one question I have at the end here:

Even if the project represents a happy path, surely there must be combinations of arguments that would cause FD to return undeclared deps (3) or unused deps (4)? For example, `--code=python_file.py --deps=requirements.txt` should cause FD to report tensorflow as undeclared, and we should expect a returncode of 3?

Reply:
About the 🤯, me too ;) This is my first serious attempt to domesticate `hypothesis`.

That is why I wanted to have a following structure:
1. dataset from which we take our samples (CLI options)
2. Atomic strategies (how do we draw from a single group of options - are they mutually exclusive, or do they have some additional constraints)
3. Chained strategy - how we actually may write a CLI arguments as users. Combination of all strategies defined in 2. We do not want to bother with order of argument groups for example.

>Even if the project represents a happy path, surely there must be combinations of arguments that would cause FD to return undeclared deps (3) or unused deps (4)? For example, --code=python_file.py --deps=requirements.txt should cause FD to report tensorflow as undeclared, and we should expect a returncode of 3?

Yes, that would also make this test more universal. I wanted to focus on creating a usable strategy first, but for the issue you mention, I thought about more general solution.
What if, for each of our sample project, we add in `expected.toml` a dictionary that assigns to each file a list of imports/dependencies that were found there. Another thing would be available packages mapping. From that we could mix and combine more tests including and excluding particular files. In case of `no_issues` tests it could look like:

```
[detailed_imports]
python_file.py: ["click", "pandas", "tensorflow"]

[detailed_requirements]
requirements.txt: ["pandas", "click"]
subdir/requirements.txt:["pandas", tensorflow"]

[packages_mapping]
pandas:["pandas"]
click:["click"]
tensorflow:["tensorflow"]

````

Reply:
Dropping this, as now test allows all non-failing exit codes.